Some editorial restructuring (#1)
soil-initiative/stacks: proposals/stacks/Overview.md

  (local.get $work)
)
```
so that when the manager resumes a thread it can do so with simply `stack.switch_call $resume $resume_thread`, which takes care of updating the global `$manager` variable, and when a thread completes it can do so with simply `stack.switch_call $complete_work $work_completed`.

Note that `$complete_work` implicitly drops the given `leafref`.

Because stack references are linear values, the stack is cleaned up by the engine once the stack frame it is stored in is cleaned up, so `$complete_work` implicitly performs this clean up before returning.
(The stack itself might contain references to other stacks within its stack frames, so this might proceed recursively.)
Unlike other memory management however, linearity enables this process to be done at a deterministic point in the execution, so systems that are sensitive to memory or timing can be ensured to have consistent behavior.

> Note that both `$resume_thread` and `$complete_work` ~~are~~ should be very simple.

Reviewer: ```suggestion
Note that both `$resume_thread` and `$complete_work` are very simple.
```

I'm confused. They are very simple. Do you mean to say that in general the intent is that `stack.switch_call` use simple calls as they are expected to be inlined?

## Summary

> This proposal outlines a suite of features that can be used to implement various patterns of non-sequential control flow: coroutines (both symmetric and asymmetric), support for asynchronous I/O, yield-style generators and many more.

Reviewer: ```suggestion
This proposal outlines a suite of features that can be used to implement various patterns of non-sequential control flow: coroutines (both symmetric and asymmetric), support for asynchronous I/O, delimited continuations and many more.
```

We don't show yield-style generators, and the common pattern for these is better done with stack inspection.

The key abstractions revolve around the concept of a stack as a manageable entity and switching control between stacks.

In order to support these, we identify two main gaps in the WebAssembly architecture: support for stack inspections and support for linearly typed variables. We explicate them here to the extent necessary for our primary use-case; but they deserve independent proposals as they have other applications.

## Appendix: Listing of new features

> The features described can be categorized into different areas: support for linear types, support for stack inspection and support for stack switching. Although stack switching depends on the former two, they are potentially separable as linear type operations and stack inspection have other applications.

Reviewer: ```suggestion
The features core to this proposal are those directly supporting first-class stacks.
However, those features rely on support for linear types and stack inspection.
Linear types and stack inspection each have utility beyond first-class stacks and so might be better factored out into separate proposals that would consider how best to design their features in a broader context.
Here we summarize the new instructions/constructs introduced above and how they fall into these three categories.
```

* `local.get_clear`

`local.get_clear` accesses a local variable and sets the variable to null after accessing the value.

```
local.get_clear $local : [] -> [t]
```

* `local.set_cleared`

`local.set_cleared` sets a local variable to the value on the stack; a trap results if the variable was not null at the start of the operation.

```
local.set_cleared $local : [t] -> []
```

* `table.get_clear`

`table.get_clear` accesses a table entry and sets the entry to null after accessing the value.

```
table.get_clear $table : [i32] -> [t]
```

* `table.set_cleared`

`table.set_cleared` sets a table entry to the value on the stack; a trap results if the table entry was not null at the start of the operation.

```
table.set_cleared $table : [i32 t] -> []
```

### Stack inspection

We list only a few of the features that we would expect from a complete suite of operations for stack inspection.

* `inquire`

`inquire` searches the stack for a handler for a given call tag and invokes that handler with the additional arguments provided.

```
inquire $call_tag : [ti*] -> [to*]
```

* `respond`

The `respond` form establishes a handler for a given call tag.

```
respond $label ins* end
```

> ### Stack Management

Reviewer: List this one first, as it's the part actually specific to first-class stacks. Come to think of it, renaming this part First-Class Stacks might make that clearer, since I think we no longer use the term "stack management" in the proposal.

Note that this proposal depends only on the exception-handling proposal (and takes advantage of some recent changes therein).

## Listing of new features

### types

* `leafRef`

* `stackRef`

### Instructions

* `inquire`

```
inquire $dispatch_tag : [ti*] -> [to*]

> ```

Reviewer: This is more of a dependency on stack inspection.

* `stack.attach_call`

* `stack.attach_clear`

```
stack.attach_clear $local
```
* `stack.attach_switch`

```
stack.attach_switch $event : [t* stackref] -> unreachable
```

* `stack.create`

```
stack.create : [] -> [stackref]
```

* `stack.extend`

```
stack.extend $rout : [t* stackref] -> [stackref]
```

* `stack.extend_leaf`

```
stack.extend_leaf $rout $event? : [ti* stackref] -> [stackref]
```

* `stack_detach`

```
stack.detach $responder $rout $label : [ti* tl*] -> unreachable
```

* `stack.start`

* `stack.switch`

```
stack.switch $event : [t* leafref] -> unreachable
```

* `stack.switch_call`

```
stack.switch_call $func $event : [ti* leafref] -> unreachable
```

* `global.get_clear`

```
global.get_clear $global : [] -> [t]
```

* `global.set_cleared`

```
global.set_cleared $global : [t] -> []
```
* `local.get_clear`

```
local.get_clear $local : [] -> [t]
```

* `local.set_cleared`

```
local.set_cleared $local : [t] -> []
```

* `table.get_clear`

```
table.get_clear $table : [i32] -> [t]
```

* `table.set_cleared`

```
table.set_cleared $table : [i32 t] -> []

> ```

Reviewer: I'd separate these as "Instructions for Linear Types" (whereas the rest are "Instructions for Stacks").

### New forms

* `rout` function

* `try redirect`

```
try instr1* redirect instr2* restore instr3* end

> ```

Reviewer: This is old.

> ```

Reviewer: Same with this.

## Editorial Notes:

* Perhaps we do not need both `stack.switch` and `stack.switch_call`. My preference would be to just have the former. My reasoning is that the toolchain will be generating this code anyway, and will likely only pick one method; and `stack.switch` is much simpler.

* There is a meta question: the extent to which code that uses stack switching has to be written for it; this includes wrapper code at places where switching may occur.

* Somewhat worried by the implicit cleanup implied by dropping a stack. Would prefer that that be arranged differently (e.g., by forcing an `abort` event on the dropped stack).

* I would definitely go for making `stack.create` an instruction. If only to reduce the cognitive burden of understanding the proposal. Suggestion: do not have a mechanism for creating attached stacks, use `stack.create` followed by attach.

* Stack (de)composition needs more TLC. Too many new/special features.

> * How does `catch $resume $resumed` work? This could use a bit more explanation.

Reviewer: Can you make this part of the pull request discussion rather than part of the overview document?
